x-pack/plugins/secret_service/server/service/secret_service.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// @ts-ignores
+import nodeCrypto from '@elastic/node-crypto';
+import crypto from 'crypto';
+import { SavedObjectsClient } from 'src/legacy/server/saved_objects';
+// @ts-ignore
+import { AuditLogger } from '../../../server/lib/audit_logger';
+
+type SSEvents = 'secret_object_created' | 'secret_object_accessed' | 'secret_object_decrypt_failed';
+type SecretId = Promise<string>;
+
+export class SecretService {
+  public readonly hideAttribute: (deets: any) => SecretId;
+  public readonly unhideAttribute: (deets: any) => any;
+  public readonly validateKey: () => Promise<boolean>;
+  private readonly type: string;
+  private readonly auditor?: AuditLogger;
+  private readonly hideAttributeWithId: (id: string, toEncrypt: any) => SecretId;
+
+  constructor(
+    savedObjectsClient: SavedObjectsClient,
+    type: string,
+    key?: Buffer,
+    auditor?: AuditLogger
+  ) {
+    this.type = type;
+    this.auditor = auditor;
+    key = key || crypto.randomBytes(128);
+    const crypt = nodeCrypto({ encryptionKey: key.toString('hex') });
+
+    const logEvent = (event: SSEvents, message?: string) => {
+      if (this.auditor) {
+        this.auditor.log(event, message);
+      }
+    };
+
+    this.hideAttribute = async (toEncrypt: any) => {
+      return this.hideAttributeWithId(crypto.randomBytes(32).toString('base64'), toEncrypt);
+    };
+
+    this.hideAttributeWithId = async (id: string, toEncrypt: any) => {
+      // now encrypt
+      const secret = await crypt.encrypt(toEncrypt);
+
+      // lastly put the encrypted details into the saved object
+      const saved = await savedObjectsClient.create(this.type, { secret }, { id });
+
+      logEvent(
+        'secret_object_created',
+        `Secret object id:${saved.id} was created at ${new Date()}`
+      );
+
+      return saved.id;
+    };
+
+    this.unhideAttribute = async (id: string) => {
+      const encKey = 'secret';
+
+      // retrieve the saved object by id.
+      const savedObject: any = await savedObjectsClient.get(this.type, id);
+
+      // extract the hidden secret value
+      const toDecrypt = savedObject.attributes[encKey];
+
+      // decrypt the details
+      try {
+        const unhidden = await crypt.decrypt(toDecrypt);
+
+        logEvent(
+          'secret_object_accessed',
+          `Saved object id:${savedObject.id} accessed at ${new Date()}`
+        );
+
+        return {
```

Comment:
Because there would be no way to reference the saved object afterwards this isn't meant to be a simple encrypt/decrypt same in same out set of methods or to be used like `(unhideAttribute(hideAttribute(X))` you would never do that. This is meant to encapsulate handling the fact that it is stored as a saved object. Perhaps the names are confusing and can be changed to better reflect their return values.

Reply:
Yeah, maybe I'm just missing the real use case example, do we have any RFC for the secret service with example?

> Perhaps the names are confusing and can be changed to better reflect their return values.

++